src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java

```diff
@@ -0,0 +1,161 @@
+                    }
+                    totalHits = new int[subScoresByQuery.length];
+                }
+                for (int i = 0; i < subScoresByQuery.length; i++) {
+                    float score = subScoresByQuery[i];
+                    // if score is 0.0 there is no hits for that sub-query
+                    if (score == 0) {
+                        continue;
+                    }
+                    totalHits[i]++;
+                    PriorityQueue<ScoreDoc> pq = compoundScores[i];
+                    ScoreDoc topDoc = pq.top();
+                    topDoc.doc = doc + docBase;
+                    topDoc.score = score;
+                    pq.updateTop();
+                }
+            }
+        };
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    protected void updateMinCompetitiveScore(Scorable scorer) throws IOException {
+        if (hitsThresholdChecker.isThresholdReached() && pqTop != null && pqTop.score != Float.NEGATIVE_INFINITY) { // -Infinity is the
+                                                                                                                    // boundary score
+            // we have multiple identical doc id and collect in doc id order, we need next float
+            float localMinScore = Math.nextUp(pqTop.score);
+            if (localMinScore > minCompetitiveScore) {
+                scorer.setMinCompetitiveScore(localMinScore);
+                totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+                minCompetitiveScore = localMinScore;
+            }
+        }
+    }
+
+    /**
+     * Get resulting collection of TopDocs for hybrid query after we ran search for each of its sub query
+     * @return
+     */
+    public TopDocs[] topDocs() {
+        TopDocs[] topDocs = new TopDocs[compoundScores.length];
+        for (int i = 0; i < compoundScores.length; i++) {
+            int qTopSize = totalHits[i];
+            TopDocs topDocsPerQuery = topDocsPerQuery(0, Math.min(qTopSize, compoundScores[i].size()), compoundScores[i], qTopSize);
+            topDocs[i] = topDocsPerQuery;
+        }
+        return topDocs;
+    }
+
+    private TopDocs topDocsPerQuery(int start, int howMany, PriorityQueue<ScoreDoc> pq, int totalHits) {
+        int size = howMany;
+
+        if (howMany < 0) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Number of hits requested must be greater than 0 but value was %d", howMany)
+            );
+        }
+
+        if (start < 0) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Expected value of starting position is between 0 and %d, got %d", size, start)
+            );
+        }
+
+        if (start >= size || howMany == 0) {
+            return EMPTY_TOPDOCS;
+        }
+
+        howMany = Math.min(size - start, howMany);
```

Comment:
I would not change value in parameter variable.
```suggestion
        int size = Math.min(howMany - start, howMany);
```

Reply:
ack